OrbitBase/include/OrbitBase/ThreadPool.h

```diff
@@ -0,0 +1,58 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef ORBIT_BASE_THREAD_POOL_H_
+#define ORBIT_BASE_THREAD_POOL_H_
+
+#include <memory>
+
+#include "OrbitBase/Action.h"
+
+// This class implements a simple fixed size thread-pool
+//
+// Usage example:
+//
+// /* Schedule an action on the thread-pool */
+// thread_poool->Schedule(CreateAction([data]{
+//   DoSomethingWith(data);
+// }));
+//
+// /* Shutdown thread-pool, blocks until all currently executed actions are
+//    complete and stops all worker threads */
+//
+// thread_pool->Shutdown();
+// thread_pool->Wait();
+//
+class ThreadPool {
+ public:
+  ThreadPool() = default;
+  virtual ~ThreadPool() = default;
+
+  // Schedules the action to be performed on the main thread.
+  virtual void Schedule(std::unique_ptr<Action> action) = 0;
+
+  template <typename F>
+  void Schedule(F&& functor) {
+    Schedule(CreateAction(std::forward<F>(functor)));
+  }
+
+  // Initiates shutdown, any Schedule after this call will fail.
+  virtual void Shutdown() = 0;
+
+  // Wait until all tasks are complete. Call this after Shutdown
+  virtual void Wait() = 0;
+
+  // Create ThreadPool with specified minimum number of worker threads
+  // and queue_threshold.
+  //
+  // When queue size hits queue_threshold thread pool creates additional
+  // thread to handle actions.
+  //
+  // TODO: If queue is empty thread_pool reduces number of worker threads
+  // until they hit thread_pool_min_size
+  static std::unique_ptr<ThreadPool> Create(size_t thread_pool_min_size,
+                                            size_t queue_threshold);
```

Comment:
I'm confused by these two parameters. The comment at the top says that this is fixed-sized, but then why is there a *min* size? Also I'm unsure of the role of `queue_threshold`.

Reply:
I clarified the comment to this method a bit. Please take a look.

Reply:
I expect thread_pool_max_size to be quite big and nearly never being reached. And yes when using for generating grpc calls it makes sense to have low threshold. I do not like default parameters in general - I would rather add a more specialized factory method.